implement onRecordEvent block [finishes #86977686]

davidsbailey opened the pull request:
implements the `onRecordEvent` block in applab as described by https://www.pivotaltracker.com/story/show/86977686

The main limitation with this approach is that it polls instead of truly listening for updates. One datapoint of the performance impact of this is that one app listening to a table with 2000 entries was able to reach our DynamoDB `development_tables` read capacity, which is currently equivalent to 40% of our capacity on `production_tables`.

Potential performance improvements include:
1. increase DynamoDB capacity
2. use Pusher to only fetch table data when something in the table has changed
3. use DynamoDB Streams / Lambda / SQS to deliver fine-grained notifications to the client, eliminating the need for the client to ever re-read the table

original design discussion (somewhat out of date, but explains potential race conditions nicely): https://docs.google.com/document/d/1MLXVhRBaRWfo_zskBiPXUkM51SS34fTfhZmvdz6wmkI/edit#heading=h.3vqoxtqv5bhp

davidsbailey commented:
Readable version of test code in the new allthethings level:

``` javascript
var lastRecordId = 0;
var recordId;
var sawCreate = false;
var sawUpdate = false;

// Find the highest record id already in the table
readRecords("mytable", {}, function(records) {
  for (var i =0; i < records.length; i++) {
    lastRecordId = Math.max(lastRecordId, records[i].id);
  }
  listenForRecords();
});


function listenForRecords() {
  onRecordEvent("mytable", function(record, eventType) {
    if (eventType === 'create') {
      if (record.id > lastRecordId) {
        // ignore records created before this test started
        handleCreateEvent(record);  
      }
    } else if (eventType === 'update') {
      handleUpdateEvent(record);
    } else if ((eventType === 'delete')) {
      handleDeleteEvent(record);
    } else {
      console.log('unexpected event type: ' + eventType);
      halt();
    }
  });
  createRecord("mytable", {name:'Alice'}, function() {});
}

function handleCreateEvent(record) {
  if (sawCreate) {
    console.log('error in handleCreateEvent: saw too many create events')
    return;
  }
  sawCreate = true;

  if (record.name !== 'Alice') {
    console.log('error in handleCreateEvent: invalid record contents: ' + JSON.stringify(record));
    return;
  }
  recordId = record.id;
  testUpdate();
}

function testUpdate() {
  updateRecord("mytable", {id:recordId, name:'Bob'}, function() {});
}

function handleUpdateEvent(record) {
  if (!sawCreate) {
    console.log('error in handleUpdateEvent: saw update before create')
    return;
  }
  if (sawUpdate) {
    console.log('error in handleUpdateEvent: saw too many update events')
    return;
  }
  sawUpdate = true;

  if (record.name !== 'Bob' || record.id !== recordId) {
    console.log('error in handleUpdateEvent: invalid record contents: ' + JSON.stringify(record));
    return;
  }
  testDelete();  
}

function testDelete() {
  deleteRecord("mytable", {id:recordId}, function() {});

}

function handleDeleteEvent(record) {
  if (!sawCreate) {
    console.log('error in handleDeleteEvent: saw delete before create')
    return;
  }
  if (!sawUpdate) {
    console.log('error in handleDeleteEvent: saw delete before update')
    return;
  }
  if (record.name !== undefined || record.id !== recordId) {
    console.log('error in handleDeleteEvent: invalid record contents: ' + JSON.stringify(record));
    return;
  }

  textLabel('successLabel', ' onRecordEvent success!');
}
```

islemaster commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,204 @@
+  /**
+   * Map from table name to table handler.
+   * @type {Object.<string, TableHandler>}
+   * @private
```
Nit: You can combine these JSDoc directives.

```
@private {Object.<string, TableHandler>}
```

islemaster commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,204 @@
+  }
+
+  if (req.status < 200 || req.status >= 300) {
+    // ignore errors
```
I assume we don't want to ignore errors here forever.  Worth a TODO to call out a need for monitoring/logging or user-notification of inability to contact the data API?

islemaster commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,204 @@
+  // the responseText check unnecessary.
+
+  var callback = this.callback_;
+  var records = JSON.parse(req.responseText);
```
This `JSON.parse` could throw an exception if something happens and the responseText isn't the valid JSON we expect.  Can we handle that gracefully?

davidsbailey replied:
I've moved the call to `scheduleNextFetch_` to before this line, so that an exception here doesn't stop polling indefinitely. Malformed JSON seems like it would represent a programming error here, so allowing this error to happen and show up in our monitoring tools seems right to me, but I am open to other ideas of how this should be handled.

islemaster replied:
Gracefully handling this path probably involves detecting and reporting it, and then acting like you didn't get a 200 status code and proceeding as such (whatever that eventually means).  Could end up looking like a retry, followed eventually by a notification to the user that certain services are currently unavailable?

islemaster commented on apps/src/applab/appStorage.js:
```diff
@@ -244,6 +246,35 @@ var handleDeleteRecord = function(tableName, record, onComplete, onError) {
+
+  if (!recordListener.addListener(tableName, onRecord)) {
+    onError('You are already listening for events on table "' + tableName + '". ' +
+      'only on event handler can be registered per table.');
```
s/on/one

islemaster commented:
What does `halt();` do in your test allthethings level?

islemaster commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,204 @@
+ * @private
+ */
+TableHandler.prototype.fetchNewRecords_ = function () {
+  var req = new XMLHttpRequest();
```
It would be nice to abstract/inject the `XMLHttpRequest` stuff to let you write unit tests around all the caching/deltas/reporting logic in here.  This has a lot in common with what we're doing in `NetSimTable`, which has a [suite of tests that fake the actual REST API](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/test/netsim/NetSimTable.js#L38).  Might be worth doing something like that here.

islemaster commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,204 @@
+      var deletedRecord = {id: parseInt(oldId, 10)};
+      callback(deletedRecord, EventType.DELETE);
+    }
+  }
```
This whole chunk of logic for detecting created/updated/deleted rows is a great candidate for unit tests.  What if you extracted this logic into a static method so you could write simple, stateless tests around it?

davidsbailey commented:
`halt()` is an undefined method... so the program halts with an error :grinning: the goal is to make sure that the "success" label doesn't get shown -- I could do it with a `success` boolean which I set (instead of `halt`) and check later if you prefer.

islemaster commented:
Ha!  That's... a little painful.  If you merge staging, now that @cpirich has finished #6927 you could actually `throw "Test failed";` if you want.

islemaster commented:
Overall, I'm a little concerned about this API that hits the callback for every touched row.

I see that there's already been some discussion around race conditions when you subscribe multiple callbacks for a single table (which I think can be solved, but more on that in a minute).  Isn't it similarly problematic to be called back for every changed row?  Suppose you wanted to do some processing on the whole table when it changed (like re-rendering a corresponding view, e.g. a sorted high score list).  Ideally you'd get called back once with a snapshot of the new table state and you'd render it.  Under this API, however, you get called back for every changed row and you re-render for every changed row - and worse, since you're only receiving the changed row in each callback, and the order of the callbacks has little to do with the order in which changes occurred, you are probably updating and rendering from a local cache of that table, and some of the intermediate states you're rendering probably never existed in the underlying table.

We may not find this causes many problems, but conceptually it worries me (and is why we stick with full table pulls in NetSim except in very specific cases).

If instead you pass the entire table, and/or make it easy/cheap to _get_ the entire table inside a callback, then the callback will always be working from a state that is at least consistent with your local cache of the remote table.  This requires that you finish updating your local cache before any callbacks, unlike your current code with does callbacks during the cache update.  Then, you may not have a problem with subscribing multiple callbacks to a table, because they'll all see the same state - no more race condition.

``` javascript
this.localCache_ = makeNewLocalCache(records);
callbacks.forEach(function (cb) {
  cb(this.localCache_);
}, this);
```

A more complex version of this requires that you still finish the cache update before any callbacks, but you queue up the individual-row callbacks in the process and make sure they also receive the full state snapshot in case they need it.

``` javascript
{this.localCache_, changes} = makeNewLocalCache(this.localCache_, records);
changes.forEach(function (change) {
  {changeType, changedRow} = change;
  callbacks.forEach(function (cb) {
    cb(changeType, changedRow, this.localCache_);
  }, this);
}, this);
```

That's all a mess -  does any of that make sense?

davidsbailey commented:
Hi Brad, thank you for the very thoughtful analysis of the API design. I had not considered the case where the entire table needs to be re-read -- the need for this block came from apps like chat (where newly-added rows could be appended to an existing blob of text), or updating an individual high score. You are right that updating a list of high scores would be hard in the model proposed in this PR. 

There are two bad scenarios to consider:
1. The API only offers incremental updates, but the app needs full table updates. The app re-reads the entire table in response to every incremental update.
2. The API only offers full table updates, but the app only needs incremental updates. The app does full table updates for each row that it wants to reread.

Both scenarios would be a bad programming experience. Both scenarios would also be terrible for our infrastructure, although caching could be used to protect our infrastructure in either case. 

My first thought is that most simple apps will benefit more from incremental updates. APIs like [Firebase](www.firebase.com) also appears to take this incremental approach. If full table updates come into high demand, then we could consider providing them as an optional last parameter to the callback (as you outlined), or alternately we could provide an optional last parameter to `onRecordEvent` that receives full table updates only once each time a group of changes are detected.

How does that sound? If you are in agreement, is there any additional work that would need to be done now?

davidsbailey commented:
I see I did not address your concern about consistency, namely that these incremental updates imply certain intermediate states which may never have existed. However I am concerned that expecting students to understand this consistency problem and having the awareness to know when to check a fully-consistent snapshot puts too much burden on the student.

Did you consider implementing fine-grained notifications within netsim? Was there a dealbreaker with that approach, or did it simply seem like it would take too long to implement?

islemaster commented:
>  or alternately we could provide an optional last parameter to onRecordEvent that receives full table updates only once each time a group of changes are detected.

I dislike this idea, since it implies more understanding of the underlying system (that you're getting updates in 'batches') than I think we want to require.

I'm much more comfortable with providing the full table state as an easily-ignored third parameter to the callback.  Of course, you could skip that and expect the student to solve/deal with the synchronization issue (which we're faced with on our layer anyway, if our underlying system goes to fine-grained notifications) but then it feels a little odd that our caching for these notifications doesn't extend to our `readTable` method, etc.  I worry that the default approach for a student who wants the whole table when it changes will be to write this code:

``` javascript
onRecordEvent("mytable", function (record, type) {
  readRecords("mytable", {}, function (allRecords) {
    // Do some work on allRecords
  });
});
```

...which of course would totally wreck our infrastructure without caching and throttling.  It makes me want to see the caching for these notifications built in such a way that it's somehow shared with the `readRecords` call.

islemaster commented:
We did end up doing incremental updates in NetSim, but they're only used on the log table which had two special properties:
1. It gets really, really big - pulling the whole thing every time over a long session created a compounding increase in the work needed to update the local cache.  All the other tables stay small.
2. We only ever INSERT to that table, never UPDATE or DELETE, so it's really safe to only do incremental updates.  We explored some capability for forcibly pulling from older rows, but in practice it's unnecessary.

The [tests for that feature](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/test/netsim/NetSimTable.js#L712) provide a pretty good overview.

In part, we ran out of time for solving fine-grained notifications; in part, NetSim does a lot of work that depends on a consistent state across the simulation (for example, there's referential integrity from wires -> nodes and from messages -> nodes).  Toward the end of that work, I also updated the cache layer for NetSim to treat `readAll` as a synchronous operation, reading the cache rather than the remote.  In that sense all sync is hidden from most of the NetSim application code, and locally it always operates on a consistent state.

That won't be so easy here, since `readTable` is already defined as an async operation, and I don't think we want to totally hide that detail from the student.

islemaster commented:
Should get @philbogle :eyes: on this if he has time.

davidsbailey commented:
This proposal sounds reasonable. To summarize:
1. update the local record cache before providing per-record `onRecordEvent` callbacks
2. ~~pass the updated local record cache as an additional last parameter 'allRecords' in the `onRecordEvent` callback~~
3. share the local record cache with `readRecords`. If the local record cache has been updated within the past second by `onRecordEvent` polling or a previous `readRecords` call, then return the locally cached records.

I think this fits well with the low-floor, high ceiling model -- we make it easier to do simple (not-necessarily-consistent) incremental updates, at the expense of making it complicated (but not impossible) to do full table updates or fully-consistent incremental updates.

@islemaster does that sound right?

@philbogle , I would love your opinion in particular on how feasible it seems to deliver in-order, fine-grained updates in the future, which would eliminate our consistency concern.

davidsbailey commented:
Dropping (2) above as a requirement per offline discussion with Brad. If we protect our infrastructure via (1) and (3), then user can call `readRecords` on every fine-grained update without overwhelming our infrastructure. If they want to optimize, they can delay/batch these calls via `setTimeout`.

davidsbailey commented:
@islemaster I've addressed all feedback. Please take another look.

islemaster commented:
:+1: Nice testing.  LGTM, expecting ongoing work on scalability.

philbogle commented:
A comment on SQS and the browser:  SQS/SNS has no browser push solution, so this doesn't really replace pusher. (You would have to poll against SQS from the browser which is not an improvement over polling against dynamo DB.)

http://stackoverflow.com/questions/9624619/is-it-possible-to-receive-an-amazon-sns-message-on-a-js-script

davidsbailey commented:
Thanks Phil, I hadn't realized SQS/SNS was a dead end. Apparently Zapier allows integration between AWS Lambda and Pusher if we were to write our own "Zapp": https://zapier.com/zapbook/aws-lambda/pusher/ (requires 14-day free trial)

@wjordan @philbogle is there any due diligence I'd need to do before connecting our AWS Lambda service to Zapier to try this out?

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+   * @private {Object.<string, TableHandler>}
+   */
+  this.tableHandlers_ = {};
+};
```
I think we want a closure surrounding the rest of the code in this module so that we don't end up polluting the global namespace (e.g. with RECORD_INTERVAL)

davidsbailey replied:
I believe that when this file gets `require`d, anything not explicitly exported will become private to this file. For example in applabTest.js I had to require RecordListener directly in order to access it, even though I was already requiring Applab which in turn requires RecordListener (which defines `var RecordListener` as a global var).

islemaster replied:
This is correct.  See [How Browserify Works#Generated closure](http://benclinkinbeard.com/posts/how-browserify-works/#generated-closure).

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+'use strict';
+
+/**
+ * Class which allows callers to listen to changes to data tables. This is currently
```
What is the rollout plan for the listener function?  

I think we probably wouldn't want to make the polling-based implementation generally available;  as we saw from Netsim even a moderate size class doing coordinated polling and fetching is enough to pose a scale risk to our servers.  Rather, we can make it available on a limited basis to get feedback on the API.

This is especially true since the implementation fetches the whole table on each poll and each table is potentially unbounded in size. Especially on mobile devices the data usage alone could be too high even if our servers don't mind.

davidsbailey replied:
I've listed remaining work items required to publicly enable and advertise this block in the description of https://github.com/code-dot-org/code-dot-org/pull/6915

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+ *
+ * The most logical incremental performance improvement is to use pusher to only
+ * re-read from the table if any of its contents have changed. Down the road, we
+ * may be able to use DynamoDB Streams / Lambda / SQS to deliver fine-grained
```
As discussed, we probably need pusher in an event to get the last mile to the browser.

davidsbailey replied:
https://www.pivotaltracker.com/story/show/114534751

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+
+/**
+ * Adds a listener which calls the callback when changes are made to the table.
+ * The callback is called once per changed record.
```
"addListener" methods typically supports multiple listeners.

Should we allow developers to add more than one listener on the same event?  This reduces coupling for larger apps because it allows components to listen for events without having to know about other components that are interested in the same event.  

If we do want to enforce the single listener, we should document that explicitly (not just in the returns clause) and perhaps call this "setListener" or "registerListener" to suggest that only a single listener is supported.

davidsbailey replied:
changed to "setListener" and improved comments. The purpose of this requirement is to warn this user if their app adds two listeners for the same table, due to the data consistency problems with adding two listeners at different times. If the need arises for a developer to use RecordListener in a different way, we can improve the API of this class to accomodate.

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+};
+
+/**
+ * Clears the timeouts and resets internal state.
```
Is this call intended to be part of the public API?  If it is, please add a description that is less implementation dependent;  the caller shouldn't have to know what you mean by "timeouts".

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+ */
+var TableHandler = function (tableName, callback) {
+  /**
+   * @private {string}
```
(This can be a one-liner if you want.)

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+ * @private
+ */
+TableHandler.prototype.handleFetchRecords_ = function (req) {
+  var done = XMLHttpRequest.DONE || 4;
```
All of our supported browsers (including IE9) support XMLHttpRequest.DONE, so I think the || 4 is unnecessary.

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+    return;
+  }
+
+  // Performance notes: on a table with 2000 records on localhost it takes ~1.5 seconds
```
1.5 seconds seems surprisingly slow, do we know where that time is going?  (Some this is probably a result of running in development mode, would be good to know the figures from test or prod)

davidsbailey replied:
When attempting to import 2000 rows I observed that production was at least 5 times faster than localhost, so I'm assuming that this number would be much lower on production.

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+ * processing the last batch of data for that table.
+ * @type {number}
+ */
+var RECORD_INTERVAL = 1000; // 1 second
```
I think polling every second is too aggressive even for an experimental launch of this feature, unless the audience is tiny and hand-picked.

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+TableHandler.prototype.fetchRecords_ = function () {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = this.handleFetchRecords_.bind(this, req);
+  var url = '/v3/shared-tables/' + Applab.channelId + '/' + this.tableName_;
```
Are there other places where we back URLs like this?  We should factor out a helper for constructing the URL to avoid duplicating knowledge of how it is formatted.

davidsbailey replied:
Adding a TODO

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+  }
+
+  if (req.status < 200 || req.status >= 300) {
+    // Ignore errors. In the future we may want to add monitoring or notify
```
It would be really good if the client reduced its polling rate in response to 5XX errors (especially 503) so that if the servers are getting overwhelmed we have a way of pushing back and reducing the polling rate.

davidsbailey replied:
https://www.pivotaltracker.com/story/show/114533591

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+
+  // Performance notes: on a table with 2000 records on localhost it takes ~1.5 seconds
+  // to fetch the data from dynamodb and 30ms to process it here. Processing time may be
+  // longer on slower machines. One quick and dirty performance improvement is to cache
```
It would be ideal for the client to specify an If-Modified-Since header on its request and have the server return 304 if there have been no changes to the table.  This would avoid sending the whole table down the wire on every request which is a large expense even if you don't parse it.

When the table is returned it would include a datestamp to use for future If-Modified-Since requests so that we don't care about inaccurate client clocks.

davidsbailey replied:
https://www.pivotaltracker.com/story/show/114533057

Finish your thoug

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+  if (req.status < 200 || req.status >= 300) {
+    // Ignore errors. In the future we may want to add monitoring or notify
+    // the user that this is happening.
+    return;
```
If we return here, I believe we don't call scheduleNextFetch, which means that polling will stop even if there is transient error.  That seems undesirable;  a transient 500 is likely to go away on the next request.

davidsbailey replied:
I've moved this check until after the call to `scheduleNextFetch_`.

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+ * @private
+ */
+TableHandler.prototype.scheduleNextFetch_ = function () {
+  this.timeoutId_ = setTimeout(function recordListener() {
```
Should we call window.setInterval instead of setTimeout so that we don't have to make repeated calls to scheduleNextFetch and run the risk of forgetting to do so?

davidsbailey replied:
I'm using setTimeout instead of setInterval for a few reasons:
-  built-in rate-limiting if the request takes a long time to fetch
- guard against requests queueing up faster than the browser (or phone!) can process them
  This approach is also slightly more amenable to varying the request rate based on receiving 503s (discussed below). This approach is a little fragile though (also discussed below). I'm going to propose reordering the code to make this as robust as possible for now, and revisiting when the 5xx rate-limiting issue below is addressed.

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+ *
+ * @constructor
+ */
+var RecordListener = module.exports = function () {
```
I think it is desirable to explicitly pass in the window dependency (needed for setTimeout etc); this makes the dependency explicit and allows for mocking.

davidsbailey replied:
I like this idea. These should be easy to extract in that way at the time that we start needing to access them from tests.

philbogle commented on apps/src/applab/RecordListener.js:
```diff
@@ -0,0 +1,230 @@
+
+TableHandler.prototype.reset = function() {
+  if (this.timeoutId_) {
+    clearTimeout(this.timeoutId_);
```
Good to make the window calls explicitly use "window.clearTimeout"

philbogle commented on apps/src/applab/api.js:
```diff
@@ -448,6 +448,13 @@ exports.deleteRecord = function (table, record, onComplete, onError) {
+  return Applab.executeCmd(null,
+                           'onRecordEvent',
+                           {'table': table,
+                            'onRecord': onRecord});
```
Just curious, what would happen if I passed the wrong type of argument to onRecord?  (Would an error be reported right away or after the event fired where it would be less clear what happened?)

philbogle commented on apps/src/applab/appStorage.js:
```diff
@@ -244,6 +246,35 @@ var handleDeleteRecord = function(tableName, record, onComplete, onError) {
+ * This is a partial implementation of onRecordEvent in the following ways:
+ * 1. it polls instead of properly listening for notifications when data changes;
+ * 2. it only issues callbacks when records are created (not updated or deleted);
+ * 3. it assumes that record ids are strictly increasing. This is not currently true but
```
Wasn't sure of implications of item 3.  Is any incorrectness introduced if record ids are not strictly increasing as is currently the case?

davidsbailey replied:
Oops, this comment was out of date. This implementation of `onRecordEvent` should be fully functional, modulo performance concerns. I've updated the comment to reflect this.

philbogle commented on apps/src/applab/appStorage.js:
```diff
@@ -244,6 +246,35 @@ var handleDeleteRecord = function(tableName, record, onComplete, onError) {
+    onError('Error listening for record events: missing required parameter "tableName"');
+    return;
+  }
+
```
isOnError required?  If not we should document; if so we should throw an error if it isn't defined.

philbogle commented on apps/src/applab/commands.js:
```diff
@@ -1566,6 +1566,13 @@ applabCommands.handleDeleteRecord = function(opts, success) {
+applabCommands.onRecordEvent = function (opts) {
+  apiValidateType(opts, 'onRecordEvent', 'table', opts.table, 'string');
+  apiValidateType(opts, 'onRecordEvent', 'callback', opts.onRecord, 'function');
+  var onError = errorHandler.handleError.bind(this, opts);
```
Do we need an apiValidateType call for onError?

davidsbailey replied:
I don't think so, because onError is not provided by the user

davidsbailey commented:
Thanks for the detailed feedback Phil. I'm working through it and adding changes and TODOs to this PR: https://github.com/code-dot-org/code-dot-org/pull/6998

islemaster commented:
FYI

```
  1 failing

  1) generateDropletModeOptions generates the expected object for applab:

      expected { Object (functions, categories) } to deeply equal { Object (functions, categories) }
      + expected - actual

               ]
             }
             "title": "onEvent"
           }
      -    "onRecordEvent": {
      -      "color": "#D3E965"
      -      "title": "onRecordEvent"
      -    }
           "penColor": {
             "color": "#4DD0E1"
             "dropdown": {
               "0": [

```

davidsbailey commented:
Thanks Brad, the test failure was real and was later fixed in https://github.com/code-dot-org/code-dot-org/commit/44d1d6ae08a5ad476bf9d92065693ede58dc444d

davidsbailey commented:
Thanks again @philbogle for your attention to this. I've finished addressing your code review feedback and have either implemented the suggestion in https://github.com/code-dot-org/code-dot-org/pull/6998 or responded to it inline here.
